### DISCUSSION TITLE
ModSecurity

### DISCUSSION THREAD

**Discussion by vdias** [dani-garcia/vaultwarden#3931] [Q&A]
Just installed the modsecurity moduel on my nginx... and getting false alerts... any one have the exclusions needed to use Vaultawarden with Modsecurity?

Found this one on my setup... cannot edit any entry using a bitwarden client:

_ModSecurity: Access denied with code 403 (phase 2). Matched "Operator `Within' with parameter `GET HEAD POST OPTIONS' against variable `REQUEST_METHOD' (Value: `PUT' ) [file "/usr/local/nginx/conf/owasp-crs/rules/REQUEST-911-METHOD-ENFORCEMENT.conf"] [line "28"] [id "911100"] [rev ""] [msg "Method is not allowed by policy"] [data "PUT"] [severity "2"] [ver "OWASP_CRS/3.3.5"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-generic"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/210/272/220/274"] [tag "PCI/12.1"] [hostname "XXXXXXXXXXX"] [uri "/api/ciphers/XXXXXXXXXXXXXXXXXXXX"] [unique_id "XXXXXXXXXXXXXXX"] [ref "v0,3"], client: XXXXXXXXXX, server: XXXXXXXXXXXXX, request: "PUT /api/ciphers/XXXXXXXXXXXXXXXX HTTP/2.0", host: "XXXXXXXXXXXXXX"_

**Comment by hboetes** [comment]
I can reproduce the issue, but this is not a bug with vaultwarden, but a bug with modsecurity.

**Reply by felbinger** [reply]
It's not a "bug". Some of the crs rules are simply triggered by vaultwarden. I'm in the process of trying to implement this too.
@vdias did you manage to create all the exceptions vaultwarden needs to run smoothly?

**Reply by morphyxz** [reply]
900200 and 920280

**Comment by morphyxz** [comment]
tldr: IDs: **900200** and **920280**

I use vaultwarden with a minimal caddy proxy (with self-signed cert - so I removed some stuff) as described here:
https://github.com/dani-garcia/vaultwarden/wiki/Using-Docker-Compose

Caddy only provides encryption in my local network.
For access on the WAN I use NPMplus (with CrowdSec, ModSecurity and CoreRuleSet).

920280 is easier. it comes AFTER crs. I added it directly to the NPMplus > Host > Advanced config:

```
modsecurity_rules 'SecRuleUpdateTargetById 920280 "!REQUEST_HEADERS:Host"';
modsecurity_rules 'SecRuleUpdateTargetById 920280 "!REQUEST_HEADERS:host"';
```

900200 comes BEFORE crs. I wasn't able to put it in the advanced settings of the host.

I put it in the **existing** "request-900-exclusion-rules-BEFORE-crs.conf" file in /opt/npmlus/modsecurity/

```
`SecAction "id:100010,phase:1,pass,nolog,setvar:tx.allowed_methods=GET HEAD POST OPTIONS PUT PATCH DELETE"`
```

**### ATTENTION DANGER!!:** This will allow those on ALL your hosts protected by this modsecure engine.

```
SecRule SERVER_ADDR "@streq 192.168.1.1" "id:100010,phase:1,pass,nolog,setvar:tx.allowed_methods=GET HEAD POST OPTIONS PUT PATCH DELETE"
SecRule SERVER_NAME "@streq any.abc.xyz" "id:100010,phase:1,pass,nolog,setvar:tx.allowed_methods=GET HEAD POST OPTIONS PUT PATCH DELETE"
```

would be much better but that didn't work for me. I need to fix the headers between the proxies I guess.

For now I just use this. I don't have any colliding domains with those URLs so it's okayish:

```
SecRule REQUEST_URI "@rx ^/(identity|api|accounts|notifications|icons|connect|admin)(/|$)" "id:100014,phase:1,pass,nolog,setvar:tx.allowed_methods=GET HEAD POST OPTIONS PUT PATCH DELETE"
```

Remember     -> ' <-     are only needed in the npmplus gui. in the .conf files you don't need those. it's a bit different but I'm sure your favourite AI will help you with that.

Have a great day!

**Comment by morphyxz** [comment]
It runs well so far but who knows if there's more.
Here is how to figure the culprits if your app breaks because of modsecurity (specify nginx error.log path):

`sudo grep "ModSecurity:" /opt/npmplus/nginx/error.log | awk -F ']' '{print $1"]", $5"]", $2"]", $7"]"}'`

If you see **Anomaly Score Exceeded** with the correct timestamp, modsecurity broke your app.

If you see:

**TX:MSC_PCRE_LIMITS_EXCEEDED**

go to modsecurity-extra.conf (or modsecurity-defaults.conf) and increase 

SecPcreMatchLimit 1000
SecPcreMatchLimitRecursion 1000

to 25000, 50000, heck, some suggest 500000. I use 25k and had no pcre_limits so far.

Now how to get the actual IDs:

On the host, create /opt/npmplus/modsecurity/debug.conf

in it:

```
SecDebugLog /data/nginx/modsec_debug.log
SecDebugLogLevel 6
SecAuditEngine RelevantOnly
SecAuditLogFormat JSON
SecAuditLog /data/nginx/modsec_audit_xyz.json
SecAuditLogParts ABDEFHIJKZ
```

Login to NPMplus GUI (or config this on whatever reverse proxy you use).
Go to the host settings of your Vaultwarden.
Under Advanced, add the following:

`modsecurity_rules_file /data/modsecurity/debug.conf`

Now RUN the process in the frontend that breaks your vaultwarden (and printed an anomaly score exceeded).
Let it break.

Right after that, take the /data/nginx/modsec_audit_xyz.json file and analyze it.
Or give it to your prefered AI to see if it can make something out of it.

EDIT: don't forget to set "SecDebugLogLevel 0" and "SecAuditEngine Off" in /opt/npmplus/modsecurity/debug.conf
(or at least do not include this file anyhwere)

Good luck hunting!
